### ISSUE TITLE
ToolTips

### ISSUE THREAD

**Issue by viaceren** [almende/vis#282]
Hi,

I've got  question about your work. I see that tooltip function is supported in 3d graph but not supported in 2d graph. What can we do about it to make tooltip function work for 2d graph?

Thank you.

**Comment by AlexDM0** [comment]
Hi,

Over the last year a lot of feature requests have been made. We have just introduced our new website which has a list of the requested features. We have placed this request on that list. 

The list can be found here:
http://visjs.org/featureRequests.html

An explaination of the new system can be found here:
http://visjs.org/blog.html#New\ website\ for\ vis\.js\!

I would like to stress that this does not mean we abandon this request. Discussion here will continue if needed on this feature but we will close it to keep our Github issue page more of a bug-todo list.

Future feature requests will still be made here and then added to the website by us.

Regards,

Alex

**Comment by rsaungc** [comment]
Hi,
Do you know when tooltop for 2D graph will be ready?  Thanks.

**Comment by AlexDM0** [comment]
Hi,

I can't give you an exact timeframe but my guess would be within the next 4 months. 

Regards,

Alex

**Comment by NotSqrt** [comment]
Is there anything we can do to help ?

This is really the last big missing feature for me.

I see that all the issues listed in the featureRequests page and related to tooltips are closed, so I don't know which issue is relevant in order to be notified of updates on this feature..

**Comment by josdejong** [comment]
Sure, feel free to implement this feature and do a pull request, it's an open source project :) 

here is a minimal guideline for contributions: https://github.com/almende/vis/blob/master/CONTRIBUTING.md

**Comment by NotSqrt** [comment]
Right.

Any heads up, since you probably had to look at this already ?

**Comment by josdejong** [comment]
What do you mean?

**Comment by NotSqrt** [comment]
You probably gave some thoughts on how to implement the hover event and the tooltips.
That would be a good starting point !

**Comment by AlexDM0** [comment]
Hi,

I have a few ideas but I'm out of the office at the moment. I'll respond to this tomorrow when I'm back. 

Regards

**Comment by NotSqrt** [comment]
Great, thanks.

**Comment by AlexDM0** [comment]
Hi,

So what I had in mind for the tooltips is the following:

Options:
className  <string>
formatFunction <function>
snapToDatapoint <boolean>
delay <number>
visible <boolean>
static <boolean>

the tooltip should be a HTML div, because that can be styled better than an SVG element, that is placed on top of the graph2d. It should be positioned next to the cursor and a line connecting it to the datapoint in question. This can be a curved line or straight. These lines are SVG elements.

The className option denotes the classname of the HTML div. If not supplied it should have a default style.

The format function takes some information on the datapoint relating to the tooltip and can customise the value being shown.

The snapToDatapoint boolean will force the tooltip to only show values on the datapoints. If it is false, it will also show the interpolated value.

The delay is in milliseconds before the tooltip pops up.

The visible argument is to have the tooltip visible or not.

The static boolean denotes if the tooltip moves along with your mouse cursor or not

We could expand the datapoints with an option that will always show the tooltip, regardless of mouse position. This means multiple tooltips can be shown at the same time.

Ideally it will work from a mouseMove event, that will give the position of the cursor which will then be referenced to all graphs, see which one is closest and pick a point. It will then emit an event with the datapoint, group and formatted value that can be used by developers to place the information in a div next to the graph instead of a tooltip.

Maybe the name tooltip becomes inaccurate at this point and we should house this under behaviour or something.

To have this high-performing and clean can be quite a challenge. You can use the binary search (if using sorted) to find the right time or timepoints and you can use the interpolation functions to get a value. The combination of SVG line connecting the datapoint and the DOM div can also be challenging.

What do you think? This is just off the top of my head so any ideas are welcome!

**Comment by hrvoj3e** [comment]
+1

**Comment by mateuszjarzewski** [comment]
What happend? Is it ready?
Im trying to implement bootstrap tooltip... Only thing I need to make it work is to vis.js allowed me to insert class, or attribute, or data-attibute...
That would be perfect
That way I cult just do it like this:

``` javascript
$(data.data)
	.each(function(index,item){
		dataset.add({
			id: item.id,
			x: item.x,
			y: item.y,
			className: 'irrelevantClassName'+item.id,
		})
	})
							})
```
And thats it, im golden.... Can I do something like this? insert class name or attribute or dataattribute or read item ID from SVG node.....
Please.....
Please...

**Comment by mateuszjarzewski** [comment]
I find a way to do this.
Basicly you just make use of item/point label:
``` javascript
dataset.add({
	x: item.x,
	y: item.y,
	label: {
		content: 'tooltip content'
	}
})
```
then one thing you can do is to hide point label like this:
``` less
.vis-line-graph {
	text {
		display: none;
	}
} 
```
or keep it by putting JSON into label
``` javascript
dataset.add({
	x: item.x,
	y: item.y,
	label: {
		content: JSON.stringify({tooltip: 'lorem', label: 'ipsum'})
	}
})
```
and then parse it while rendering...

All this is overcomplicated but require no core patching, so it's the best...
If more explaining is needed, please make a comment...

Trala.

**Comment by mateuszjarzewski** [comment]
I find a way to do this.
Basically I just make use of item/point label
``` javascript
dataset.add({
	x: item.x,
	y: item.y,
	label: {
		content: 'tooltip content'
	}
})
```
One thing You can do, is to hide point label
``` less
.vis-line-graph {
	text {
		display: none;
	}
} 
```
or retain it by using JSON in label
``` javascript
dataset.add({
	x: item.x,
	y: item.y,
	label: {
		content: JSON.stringify({tooltip: 'lorem', label: 'ipsum'})
	}
})
```
and then parse it while rendering...

All this is overcomplicated but require no core patching, so it's the best...
If something's not clear or require more explaining, please make a comment...

Trala.

**Comment by mateuszjarzewski** [comment]
I find a way to do this.
Basically I just make use of item/point label
``` javascript
dataset.add({
	x: '31',
	y: '337',
	label: {
		content: 'tooltip content'
	}
})
```
One thing You can do, is to hide point label
``` less
.vis-line-graph {
	text {
		display: none;
	}
} 
```
or retain it by using JSON in label
``` javascript
dataset.add({
	x: '31',
	y: '337',
	label: {
		content: JSON.stringify({tooltip: 'lorem', label: 'ipsum'})
	}
})
```
and then parse it while rendering...

All this is overcomplicated but require no core patching, so it's the best...
If something's not clear or require more explaining, please make a comment...

Trala.

**Comment by siben168** [comment]
@mateuszjarzewski 
Could you explain how you stick your tooltip content with bar chart or something? I tried your solution with bootstrap tooltip but have on idea how to show up tooltip with Json.

A sample code you are using now will be good enough. 

Thanks
Eric

**Comment by nickimola** [comment]
Yes please, If you can provide a code example it will be really nice :)

**Comment by mojoaxel** [comment]
We should provide an example that shows how to use tooltips.

**Comment by sunsetjunks** [comment]
Hello! graph2d has a wonderful performance, sampling, etc. But I really miss the basic tooltip functionality... 

When could we get it working? Thanks!

**Comment by jgibson** [comment]
I needed basic tooltips (and select, see #311) on my graph so I took a crack at implementing this and ran into a few issues. @mdxs' suggestion worked pretty well. I found that a `title` attribute didn't work, but that an embedded `title` element did. The second was more tedious. As an optimization vis.js reuses SVG elements after altering the graph or scrolling a bunch. So the code has to be careful to reset existing titles lest points wind up with incorrect titles.

I also took @mateuszjarzewski's advice and tried a 3rd party tooltip library ([qtip2](http://qtip2.com/)) and it faced the same issue, only instead of clearing out the title element I had to deal with stale event handlers.

However I found a larger problem with both approaches when dealing with multiple groups on the graph. I found that paths from groups that were rendered later would obscure the tooltips on dots from the groups that were rendered earlier. Both Safari and FF were using the bounding box of a path to detect mouse events and thus a single path would grab all of the events instead of any points. To get around the issue I ended up changing the rendering code to render all of the paths before any of the points.

I also explored using [SVGSVGElement.getIntersectionList()](https://developer.mozilla.org/en-US/docs/Web/API/SVGSVGElement) to handle the mouse events and determine what was being clicked or hovered over. However I found that Firefox simply doesn't support the feature (as of February 2017) and that Safari was matching via bounding box instead of by the actual geometry. I also don't know what the performance impact would be to use this for tooltips given that you'd be continually firing this event as you interacted with the graph. If the goal is to implement @AlexDM0's extensive requirements then I'd guess that this would be the correct approach, but it could be painful to get it work fast and well, especially given the current level of browser support.

**Comment by jgorene** [comment]
Hello guys,
Take a look [on this track.](http://bl.ocks.org/d3noob/a22c42db65eb00d4e369).. very easy to achieve for a really nice result (in my case)
I also think that vis.js is a awesome library and tooltip such as found onTimeline is missing for graph2d. Unfortunately, I'm not still rather quite expert to implement this feature ;)
Anyway, I often use d3.js a nice library very complementary to vis.js and if this track can be useful to someone else, for now .....

**Comment by Nokel81** [comment]
Is it possible to know a date when this will become available even in beta form?

**Comment by AnushaputtaUI** [comment]
is there a tooltip on graph 2D now if not,Is there any solution?
